### DISCUSSION TITLE
Can't set a json default value on MySQL

### DISCUSSION THREAD

**Discussion by hcoura** [mikro-orm/mikro-orm#4926] [General]
**Describe the bug**
After mysql 8.0.13 we can set DEFAULT values for JSON columns like

`DEFAULT ('{a: 1}')`

We can't do this via the `default` option from the `@Property` but we can via `defaultRaw` something like:

```
  @Property({
    type: 'json',
    defaultRaw: `('${JSON.stringify({
      orderTextUpdates: false,
      driver: NotificationDriverSettingEnum.TextAndEmail,
    })}')`,
  })
  notificationSettings: NotificationSettingsType;
```

will generate this schema change

```
import { Migration } from '@mikro-orm/migrations';

export class Migration20231114174123_AddsPatientNotificationSettings extends Migration {
  async up(): Promise<void> {
    this.addSql(
      'alter table `patients` add `notification_settings` json not null default  (\'{"orderTextUpdates":false,"driver":"text_and_email"}\');',
    );
  }

  async down(): Promise<void> {
    this.addSql('alter table `patients` drop `notification_settings`;');
  }
}
```

All good with mysql and works as expected, but if we generate a schema diff again we will get:

```
import { Migration } from '@mikro-orm/migrations';

export class Migration20231114174809_AddsPatientNotificationSettings extends Migration {
  async up(): Promise<void> {
    this.addSql(
      'alter table `patients` modify `notification_settings` json not null default (\'{"orderTextUpdates":false,"driver":"text_and_email"}\');',
    );
  }

  async down(): Promise<void> {
    this.addSql(
      'alter table `patients` modify `notification_settings` json not null default _utf8mb4\'\'{"orderTextUpdates":false,"driver":"text_and_email"}\'\';',
    );
  }
}
```

The changes from https://github.com/mikro-orm/mikro-orm/pull/3503 won't help because this is on the `default/defaultRaw` fields.

**Versions**

| Dependency | Version |
| - | - |
| node | 20.5.0 |
| typescript | 5.1.3 |
| mikro-orm | 5.9.3 |
| your-driver | mysql |

**Comment by B4nan** [comment]
Why do you have the parens around the JSON, is that required for the query to work? My guess is that's the problem. You need to provide the exact same value as what gets inferred from the database with `defaultRaw`.

Try using `MIKRO_ORM_VERBOSE` env var when running the CLI, it will print _a lot_ of debug info, including why the values are considered different.

```bash
# you can use schema:update for testing this, its what the migrator uses under the hood
MIKRO_ORM_VERBOSE=1 npx mikro-orm schema:update -d
```

**Reply by hcoura** [reply]
Well the parans are needed because MySql requires then:

https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html#data-type-defaults.html#data-type-defaults-explicit

> The [BLOB](https://dev.mysql.com/doc/refman/8.0/en/blob.html), [TEXT](https://dev.mysql.com/doc/refman/8.0/en/blob.html), GEOMETRY, and [JSON](https://dev.mysql.com/doc/refman/8.0/en/json.html) data types can be assigned a default value only if the value is written as an expression, even if the expression value is a literal:

`My guess is that's the problem.` 

That's my guess as well, but we can't set the defaultRaw to the value inferred from the database as it's required it is an expression and mysql will transform that it what we see later.

The issue looks more complex than something I could workaround.

**Reply by B4nan** [reply]
I guess we could trim the parens automatically for JSON columns, sounds safe as it's not a valid JSON with the parens anyway. Can you confirm it helps if you remove them? It should no longer generate the diff, you should see the values inferred form the database with the verbose flag enabled.

**Reply by hcoura** [reply]
Not sure I did what you want, but after removing the parens in the `defaultRaw` and running `MIKRO_ORM_VERBOSE=1 npx mikro-orm schema:update -d` gave these schema logs

```
[schema] 'default' changed for column patients.notification_settings { column1: { name: 'notification_settings', type: 'json', mappedType: JsonType {}, unsigned: false, length: null, default: `_utf8mb4\\'{"orderTextUpdates":false,"driver":"text_and_email"}\\'`, nullable: true, primary: false, unique: false, autoincrement: false, precision: null, scale: null, comment: '', extra: '', enumItems: [] }, column2: { name: 'notification_settings', type: 'json', mappedType: JsonType {}, unsigned: false, autoincrement: false, primary: false, nullable: true, default: `'{"orderTextUpdates":false,"driver":"text_and_email"}'` }}
[schema] column patients.notification_settings changed { changedProperties: Set(1) { 'default' } }
```


And generating a migration gave:

```
export class Migration20231114184222_AddsPatientNotificationSettings extends Migration {
  async up(): Promise<void> {
    this.addSql(
      'alter table `patients` modify `notification_settings` json default \'{"orderTextUpdates":false,"driver":"text_and_email"}\';',
    );
  }

  async down(): Promise<void> {
    this.addSql(
      'alter table `patients` modify `notification_settings` json default _utf8mb4\'\'{"orderTextUpdates":false,"driver":"text_and_email"}\'\';',
    );
  }
}

```

**Reply by B4nan** [reply]
What if you do this?

```ts
  @Property({
    type: 'json',
    defaultRaw: `_utf8mb4'{"orderTextUpdates":false,"driver":"text_and_email"}'`,
  })
  notificationSettings: NotificationSettingsType;
```

(you might have to adjust the escaping, this is just a copy paste, tried to clean it up a bit but idk)

**Reply by hcoura** [reply]
If I do 
```
defaultRaw: `_utf8mb4\\\'{"orderTextUpdates":false,"driver":"text_and_email"}\\\'`,
```

Mikro won't see the diff anymore, but this would only "work" if I set the migration manually to something like

```
this.addSql(
      'alter table `patients` add `notification_settings` json null default (\'{"orderTextUpdates":false,"driver":"text_and_email"}\');',
    );
```

That is, if I generate the migration with that `defaultRaw` value mysql will throw a syntax error

**Reply by B4nan** [reply]
Got it, will try to look into this later today and hopefully make it work with what you tried initially, it should be possible to strip both the `_utf8mb4` thing as well as the parens from what is inferred from the schema.

**Reply by B4nan** [reply]
this will help https://github.com/mikro-orm/mikro-orm/commit/53c39118851eac1ad10e1f32d9e88055a4476018?w=1

**Reply by hcoura** [reply]
Awesome! Thanks a lot!

**Comment by hcoura** [comment]
@B4nan Your fix/release does work in the sense that it doesn't detect the change anymore.

That said, the diff is still there somewhere haha, today I made some unrelated change to the schema and when I generate a new schema I get this on the migration

```
  async down(): Promise<void> {
    // ...
    this.addSql('alter table `patients` modify `notification_settings` json not null default _utf8mb4\'\'{"orderTextUpdates":false,"driver":"text_and_email"}\'\';');
  }
```

Interestingly it's only on `down` and if I remove it and run the migration and try to generate again no differences will be detected.

**Reply by B4nan** [reply]
> Your fix/release does work in the sense that it doesn't detect the change anymore.

But what change? It won't detect changes in whitespaces, I'd say that's a good enough solution.

edit: maybe I misunderstood what you are trying to say. Anyway, repro welcome, I can't fix things based on a generated migration.

**Reply by hcoura** [reply]
I will try to build a simple repro later today or tomorrow, but what I am trying to say is that for some unrelated change, the schema generator still thinks something should happen on `down`.

**Reply by B4nan** [reply]
Oh right, got that reproduced in the test I made already.

**Reply by B4nan** [reply]
cccc7561f9dc3c90f22ad293fceb7b58a67dabea should help
